Enable limit on range Of JSON-RPC API trace_filter method (#5971)

alyokaz opened the pull request:
Enable a limit on the range of blocks that can be supplied to the JSON-RPC trace_filter method.

The limit has a default value and can be overridden with a command line option at start up.

fixes #5971

Gabriel-Trintinalia commented on besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:
```diff
@@ -1259,6 +1259,12 @@ static class PermissionsOptionGroup {
+  @Option(
+      names = {"--rpc-max-trace-range"},
+      description =
+          "Specifies the maximum number of blocks for the trace_filter method (default: $DEFAULT-VALUE)")
```
```suggestion
          "Specifies the maximum number of blocks for the trace_filter method. Must be >=0. 0 specifies no limit (default: $DEFAULT-VALUE)")
```

alyokaz replied:
Done

Gabriel-Trintinalia commented on besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:
```diff
@@ -1259,6 +1259,12 @@ static class PermissionsOptionGroup {
   private final Integer numberOfblocksToCache = 0;
 
+  @Option(
+      names = {"--rpc-max-trace-range"},
```
I think we can renamed it to `--rpc-max-trace-filter-range` as it only applies for this method.

alyokaz replied:
Done

Gabriel-Trintinalia commented on besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:
```diff
@@ -2526,7 +2532,8 @@ private ApiConfiguration apiConfiguration() {
             .gasCap(rpcGasCap)
-            .isGasAndPriorityFeeLimitingEnabled(apiGasAndPriorityFeeLimitingEnabled);
+            .isGasAndPriorityFeeLimitingEnabled(apiGasAndPriorityFeeLimitingEnabled)
+            .maxTraceRange(maxTraceRange);
```
same here, maybe maxTraceFilterRange

alyokaz replied:
Done

Gabriel-Trintinalia commented on ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilter.java:
```diff
@@ -88,6 +93,19 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
     final long toBlock = resolveBlockNumber(filterParameter.getToBlock());
     LOG.trace("Received RPC rpcName={} fromBlock={} toBlock={}", getName(), fromBlock, toBlock);
 
+    try {
+      if (maxRange > 0 && toBlock - fromBlock > maxRange)
+        throw new IllegalArgumentException(RpcErrorType.EXCEEDS_RPC_MAX_BLOCK_RANGE.getMessage());
+    } catch (IllegalArgumentException ex) {
+      LOG.atDebug()
+          .setMessage("eth_getLogs request {} failed:")
+          .addArgument(requestContext.getRequest())
+          .setCause(ex.getCause())
+          .log();
+      return new JsonRpcErrorResponse(
+          requestContext.getRequest().getId(), RpcErrorType.EXCEEDS_RPC_MAX_BLOCK_RANGE);
+    }
+
```
we could remove the try-catch here and use a `if` instead

alyokaz replied:
Done

Gabriel-Trintinalia commented on besu/src/test/resources/everything_config.toml:
```diff
@@ -90,6 +90,7 @@ rpc-max-logs-range=100
 json-pretty-print-enabled=false
 cache-last-blocks=512
 rpc-gas-cap = 50000000
+rpc-max-trace-range=100
```
same as above

alyokaz replied:
Done

Gabriel-Trintinalia reviewed (commented):
Some comments in the code. 

- This is also a breaking change. Could we add a entry to the changelog? 
- Also, a unit test similar to `shouldFailIfParamsExceedMaxRange` would be nice

alyokaz commented on ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilter.java:
```diff
@@ -88,6 +93,19 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+        throw new IllegalArgumentException(RpcErrorType.EXCEEDS_RPC_MAX_BLOCK_RANGE.getMessage());
+    } catch (IllegalArgumentException ex) {
+      LOG.atDebug()
+          .setMessage("eth_getLogs request {} failed:")
```
Done

alyokaz commented:
> Thank you for the PR! I have added some comments in the code.
> 
> * This is a breaking change. Could you please add a entry to the changelog?
> * Also, a unit test similar to `shouldFailIfParamsExceedMaxRange` would be nice

Changelog entry added.

I've added a unit test for TraceFilter covering the new option. I wanted to add a test for the `0` parameter, but without mocking `traceFilterWithPipeline` using something like Powermock, it would require extensive refactoring and I felt it might run orthogonal to the "big function" style of the code. I think maybe it might be better dealt with in an integration test?

Gabriel-Trintinalia reviewed (commented):
Looking good, just a text suggestion and the the right copyright for the test class

alyokaz commented on CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ## 24.1.2-SNAPSHOT
 
 ### Breaking Changes
+- A Limit on the range of blocks passed to the JSON-RPC API `trace-filter` method, that defaults to 1000 and can be set with `--rpc-max-logs-range`, has been enabled [#6446](https://github.com/hyperledger/besu/pull/6446)
```
Changed as suggested

alyokaz commented on ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilterTest.java:
```diff
@@ -0,0 +1,66 @@
+/* SPDX-License-Identifier: Apache-2.0 */
```
Added the copyright as it appears at the top of EthGetLogsTest

macfarla commented on CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ## 24.1.2-SNAPSHOT
 
 ### Breaking Changes
+- The `trace-filter` method in JSON-RPC API now has a default block range limit of 1000, adjustable with `--rpc-max-logs-range` [#6446](https://github.com/hyperledger/besu/pull/6446)
```
```suggestion
- The `trace-filter` method in JSON-RPC API now has a default block range limit of 1000, adjustable with `--rpc-max-trace-filter-range` [#6446](https://github.com/hyperledger/besu/pull/6446)
```

macfarla commented on ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilterTest.java:
```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright ConsenSys AG.
```
```suggestion
 * Copyright Hyperledger Besu Contributors.
```

macfarla reviewed (approved):
approved with a couple of minor nits
